### PR TITLE
Add blackspot frontend

### PR DESCRIPTION
**Pull request by Lknechtli**
Add layer to windshaft
Add blackspot resource service
Style based on severity
Only show layer as option if blackspots exist
Show blackspots in dashboard minimap
Show blackspots layer on main map

### CODE REVIEW

File: web/app/scripts/black-spots/black-spots-directive.js [hunk 1]
```diff
@@ -1,20 +1,90 @@
...
+                    }
+                    // If it's already in the right format, don't do the conversion
+                    else if (dtString.indexOf('/') <= 0) {
+                        maxDateString = dtString + 'T23:59:59Z';
+                    } else {
+                        var components = dtString.split('/');
+                        var month = components[0];
+                        var day = components[1];
+                        var year = components[2];
+                        maxDateString = year + '-' + month + '-' + day + 'T23:59:59Z';
```
Comment (Lknechtli) [inline web/app/scripts/black-spots/black-spots-directive.js]: Similar logic exists elsewhere - would it be worth refactoring out into a utility or something?

Comment (kshepard) [inline web/app/scripts/black-spots/black-spots-directive.js]: Yes, refactoring the common logic would be great.

File: web/app/scripts/black-spots/black-spots-directive.js [hunk 2]
```diff
@@ -1,20 +1,90 @@
...
+                    }
+                }
+
+                $q.all([leafletController.getMap(),
+                        BlackspotSets.query({
+                            'effective_at': maxDateString
```
Comment (kshepard) [inline web/app/scripts/black-spots/black-spots-directive.js]: This needs a record_type parameter as well, so we get the appropriate black spots.

File: web/app/scripts/views/map/layers-controller.js [hunk 3]
```diff
@@ -259,13 +259,42 @@
...
+                    var day = components[1];
+                    var year = components[2];
+                    maxDateString = year + '-' + month + '-' + day + 'T23:59:59Z';
+                }
+            }
             $q.all([TileUrlService.recTilesUrl(ctl.recordType),
                     TileUrlService.recUtfGridTilesUrl(ctl.recordType),
-                    TileUrlService.recHeatmapUrl(ctl.recordType)]).then(function(tileUrls) {
+                    TileUrlService.recHeatmapUrl(ctl.recordType),
+                    BlackspotSets.query({'effective_at': maxDateString}).$promise
```
Comment (kshepard) [inline web/app/scripts/views/map/layers-controller.js]: Also add the record_type here.

File: web/app/scripts/views/map/layers-controller.js [hunk 4]
```diff
@@ -259,13 +259,42 @@
...
+                    .then(function(blackspotSet){
+                        var set = blackspotSet[blackspotSet.length-1];
```
Comment (kshepard) [inline web/app/scripts/views/map/layers-controller.js]: Spaces between operator.

File: web/app/scripts/views/map/layers-controller.js [hunk 5]
```diff
@@ -259,13 +259,42 @@
...
+                        if(set){
```
Comment (kshepard) [inline web/app/scripts/views/map/layers-controller.js]: Space after `if` (here, and a few other places in this PR).

File: windshaft/driver.js [hunk 6]
```diff
@@ -128,11 +149,21 @@ function setRequestParameters(request, callback, redisClient) {
...
-                + endBoundaryQuery;
+            params.sql = baseBoundaryQuery + filterBoundaryQuery +
+                params.id + "'" + endBoundaryQuery;
         }
 
         callback(null, request);
+    } else if (params.tablename === 'black_spots_blackspot') {
+        //record type, filter effective at
+        params.style = blackspotStyle;
+        if (params.id === 'ALL') {
```
Comment (kshepard) [inline windshaft/driver.js]: This `ALL` logic should be removed here.

File: web/app/scripts/black-spots/black-spots-directive.js [hunk 7]
```diff
@@ -1,20 +1,90 @@
...
         };
         return module;
+
+        function link(scope, element, attrs, controllers) {
+            InitialState.ready().then(function() {
+                var leafletController = controllers[0];
+
+                leafletController.getMap().then(addBaseLayers);
+                var maxDateString;
+                if (FilterState.filters.hasOwnProperty('__dateRange')) {
```
Comment (kshepard) [inline web/app/scripts/black-spots/black-spots-directive.js]: If `__dateRange` doesn't exist in the filters, the query should still be sent, but with the current datetime as `effective_at`. If the user hasn't manually edited the date range filter yet (e.g. the first time they load the page), this variable won't be set, but we should still show the current black spots.

**Comment by kshepard** [comment]
Looks good! The tests are failing, so those need to be fixed. New tests should be added too (at least a basic one for the black spots service. +1 after that, and after addressing my few comments.

**Comment by Lknechtli** [comment]
- [ ] Fix tests
- [ ] Refactor max time from filter into the service

File: web/app/scripts/black-spots/black-spots-controller.js [hunk 8]
```diff
@@ -1,11 +1,73 @@
...
     /* ngInject */
-    function BlackSpotsController() {
+    function BlackSpotsController(
+        InitialState, TileUrlService, FilterState, RecordState, BlackspotSets
+    ) {
+        var cartoDBAttribution = '&copy; <a href="http://www.openstreetmap.org/copyright">OpenStreetMap</a> contributors, &copy; <a href="http://cartodb.com/attributions">CartoDB</a>';
+        var ctl = this;
+        ctl.map = null;
+        ctl.initMap = initMap;
+        initialize();
```
Comment (kshepard) [inline web/app/scripts/black-spots/black-spots-controller.js]: If there's no code being run on initialize, this call and the function can be removed.

File: web/app/scripts/map-layers/tile-url-service.js [hunk 9]
```diff
@@ -39,6 +42,11 @@
             return _insertIdAtALL(allBoundariesUrl, boundsUuid);
         }
 
+        function blackspotTilesUrl(blackspotSet) {
+            return _insertIdAtALL(blackspotsUrl, blackspotSet);
+
```
Comment (kshepard) [inline web/app/scripts/map-layers/tile-url-service.js]: Extra newline here.

File: deployment/ansible/roles/driver.web/templates/index.html.j2 [hunk 10]
```diff
@@ -107,6 +107,7 @@
     <script src="scripts/resources/query-builder-service.js"></script>
     <script src="scripts/resources/aggregate-query-service.js"></script>
     <script src="scripts/resources/saved-filters-service.js"></script>
+    <script src="scripts/resources/blackspots-service.js"></script>
```
Comment (kshepard) [inline deployment/ansible/roles/driver.web/templates/index.html.j2]: I think it's worth adding a basic test for this module for coverage reasons/verifying it loads. All of the resource tests are really straightforward, so it'll mostly be a direct copy/paste from one of the others.

**Comment by kshepard** [comment]
+1 after adding blackspots-service test.

**Pull request merged**